pkg/status/workstatus.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const workStatusIdentificationIndexKey = "workStatusIdentificationIndex"
+
+type workStatus struct {
+	wecName                string
+	sourceObjectIdentifier util.ObjectIdentifier
+	status                 map[string]interface{}
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, key string) error {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
```

Comment:
FYI: If instead you had defined `type workStatusRef cache.ObjectName` then there would be no need to call a parsing function here and no need for error handling here.

Not a blocker.

Reply:
The current WorkStatus name is not safe to work with for inferring source object / wecname in my opinion. Therefore I changed this ref to the portion of the `workStatus` struct that deals with bookkeeping.

Reply:
@waltforme

Reply:
Remember that we are doing eventually consistent programming with no timing guarantees in a system of components that can individually go up and down. We should assume no limit (apart from what we explicitly program) on what other goroutines and other processes do between two consecutive memory operations in one goroutine.

How much should this package assume about the behavior of code in the ks/OSA repo? How much should this package assume about clients manipulating an ITS in general?

Do we expect the `.spec.sourceRef` of a WorkStatus to be immutable?

It is not immutable today. The IDL does not say that it is immutable. We do not lock down the authority to make changes: the examples, documentation, and tests make liberal use of super-user credentials.

Even if the `.spec.sourceRef` of a WorkStatus were immutable, that would not stop one client from deleting a WorkStatus and then creating a replacement with the same name but different `.spec.sourceRef` (and the status controller adds no finalizer that would prevent that from happening without interlock with the controller). Informers are only eventually consistent; they do not promise to deliver a delete and a create event in this case; the informer client might only see an update event.

If the `.spec.sourceRef` of a WorkStatus _were_ immutable then I think that we could consider it to be an extended part of the object's identity. If we did, then a Get by name would have to be followed by checking the `.spec.sourceRef` and, on mismatch with the expected SourceRef, acting like the Get returned NotFound.

To properly handle `.spec.sourceRef` of a WorkStatus being mutable, the status controller should keep track of what it last got from each WorkStatus --- so that upon change in `.spec.sourceRef` the data previously associated with that WorkStatus can be cleared out.

Reply:
Will be addressed in followup PR - logged in issue #2141.